### PR TITLE
[DataObjects] Allow multiple assignments on advanced (object) relations

### LINKED ISSUE #2022: Object relations - allow assignment of an item multiple times
# Feature Request 

Currently the relation object types (e.g. Multi-Href, Multi-Href with Metadata ) only allow the assignment of an item once. It would be great if we can assign one item multiple times.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #2022

### CODE REVIEW

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    public function up(Schema $schema)
+    {
+        try {
+            $db = Db::get();
+
+            // add index to primary key in relations tables
+            $relationTables = $db->fetchAll("SHOW TABLES LIKE 'object\_relations\_%'");
+            foreach ($relationTables as $table) {
+                $relationTable = current($table);
+                $this->addSql('ALTER TABLE `' . $relationTable . '` ADD PRIMARY KEY `PRIMARYKEY` (`src_id`, `dest_id`, `ownertype`, `ownername`, `fieldname`, `type`, `position`, `index`), DROP INDEX `PRIMARY`');
```
Comment (fashxp) [inline bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: why not drop primary key first and then create new primary key with same name? something like that: 

``` 
alter table object_relations_33 drop primary key;

alter table object_relations_33
	add primary key (src_id, dest_id, ownertype, ownername, fieldname, type, position, `index`);
```

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            }
+
+            $metaTables = $db->fetchAll("SHOW TABLES LIKE 'object\_metadata\_%'");
+            foreach ($metaTables as $table) {
+                $metaTable = current($table);
+
+                // add index column to metadata tables
+                $this->addSql('ALTER TABLE `' . $metaTable . '` ADD COLUMN `index` int(11) unsigned NOT NULL DEFAULT \'0\'');
+                $this->addSql('ALTER TABLE `' . $metaTable . '` ADD PRIMARY KEY `PRIMARYKEY` (`o_id`, `dest_id`, `type`, `fieldname`, `column`, `ownertype`, `ownername`, `position`, `index`), DROP INDEX `PRIMARY`');
```
Comment (fashxp) [inline bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: see comment above

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    public function down(Schema $schema)
+    {
+        try {
+            $db = Db::get();
+            // remove index from primary key in relations tables
+            $relationTables = $db->fetchAll("SHOW TABLES LIKE 'object\_relations\_%'");
+            foreach ($relationTables as $table) {
+                $relationTable = current($table);
+                $this->addSql('ALTER TABLE `' . $relationTable . '` ADD PRIMARY KEY `PRIMARYKEY` (`src_id`, `dest_id`, `ownertype`, `ownername`, `fieldname`, `type`, `position`), DROP INDEX `PRIMARY`');
```
Comment (fashxp) [inline bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: see comment above. 
in addition to that, this might fail since there might be data where already multiple assignments took place

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                        ON  mt.o_id = rl.src_id
+                            AND mt.dest_id = rl.dest_id
+                            AND mt.type = rl.type
+                            AND mt.fieldname = rl.fieldname
+                            AND mt.ownertype = rl.ownertype
+                            AND mt.ownername = rl.ownername
+                            AND mt.position = rl.position
+                        SET mt.index = rl.index');
+            }
+        } catch (\Exception $e) {
```
Comment (fashxp) [inline bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: never ever catch exceptions and to nothing!

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            }
+
+            // drop index column from metadata tables
+            $metaTables = $db->fetchAll("SHOW TABLES LIKE 'object\_metadata\_%'");
+            foreach ($metaTables as $table) {
+                $metaTable = current($table);
+                $this->addSql('ALTER TABLE `' . $metaTable . '` DROP COLUMN `index`');
+                $this->addSql('ALTER TABLE `' . $metaTable . '` ADD PRIMARY KEY `PRIMARYKEY` (`o_id`, `dest_id`, `type`, `fieldname`, `column`, `ownertype`, `ownername`, `position`), DROP INDEX `PRIMARY`');
+            }
+        } catch (\Exception $e) {
```
Comment (fashxp) [inline bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: never ever catch exceptions and to nothing!

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js [hunk 6]
```diff
@@ -856,7 +857,7 @@ pimcore.object.tags.manyToManyObjectRelation = Class.create(pimcore.object.tags.
                 var key;
 
                 if (rdata.success) {
-                    var rec = this.store.getById(item.id);
+                    var rec = newItem[0];
```
Comment (fashxp) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js]: why did you change that line?

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js [hunk 7]
```diff
@@ -669,21 +670,35 @@ pimcore.object.tags.manyToManyRelation = Class.create(pimcore.object.tags.abstra
...
 
-    requestNicePathData: function (targets) {
+    requestNicePathData: function(targets, isInitialLoad) {
         if (!this.object) {
             return;
         }
+
+        var context = this.getContext();
+        var loadEditModeData = false;
+        if(isInitialLoad && this.fieldConfig.optimizedAdminLoading && context['containerType'] == 'object') {
```
Comment (fashxp) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js]: why did you add that here? optimizedAdminLoading is never available for manyToManyRelations

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js (hunk 6 above)
Comment (dvesh3) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyObjectRelation.js]: because id is not unique anymore!

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js (hunk 7 above)
Comment (dvesh3) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/tags/manyToManyRelation.js]: good catch...I'll remove it ;)

File: bundles/CoreBundle/Resources/migrations/Version20190403120728.php (hunk 3 above)
Comment (dvesh3) [reply bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: I agree...but still there is chance of downgrade without changing the data! Do you think it is better to leave the index as is?

Comment (fashxp) [reply bundles/CoreBundle/Resources/migrations/Version20190403120728.php]: I think we should try and throw an error, if it is not possible - with hints, what needs to be done.
